File: TypescriptWebSplatPort/web-splat-port/ts-web-splat/src/controller.ts
import { vec3, quat, mat3 } from "gl-matrix";
import { PerspectiveCamera } from "./camera.js";

export class CameraController {
  public center: vec3 = vec3.create();
  public up: vec3 | null = null;
  private amount: vec3 = vec3.create();
  private shift: [number, number] = [0, 0];
  private rotation: vec3 = vec3.create();
  private scroll: number = 0;
  public speed: number;
  public sensitivity: number;

  public leftMousePressed: boolean = false;
  public rightMousePressed: boolean = false;
  public altPressed: boolean = false;
  public userInput: boolean = false;

  constructor(speed: number, sensitivity: number) {
    this.speed = speed;
    this.sensitivity = sensitivity;
  }

  processKeyboard(key: string, pressed: boolean): boolean {
    const amount = pressed ? 1.0 : 0.0;
    let processed = false;

    switch (key) {
      case "KeyW":
      case "ArrowUp":
        this.amount[2] += amount;
        processed = true;
        break;
      case "KeyS":
      case "ArrowDown":
        this.amount[2] += -amount;
        processed = true;
        break;
      case "KeyA":
      case "ArrowLeft":
        this.amount[0] += -amount;
        processed = true;
        break;
      case "KeyD":
      case "ArrowRight":
        this.amount[0] += amount;
        processed = true;
        break;
      case "KeyQ":
        this.rotation[2] += amount / this.sensitivity;
        processed = true;
        break;
      case "KeyE":
        this.rotation[2] += -amount / this.sensitivity;
        processed = true;
        break;
      case "Space":
        this.amount[1] += amount;
        processed = true;
        break;
      case "ShiftLeft":
        this.amount[1] += -amount;
        processed = true;
        break;
    }

    this.userInput = processed;
    return processed;
  }

  processMouse(mouseDx: number, mouseDy: number): void {
    if (this.leftMousePressed) {
      this.rotation[0] += mouseDx;
      this.rotation[1] += mouseDy;
      this.userInput = true;
    }
    if (this.rightMousePressed) {
      this.shift[1] += -mouseDx;
      this.shift[0] += mouseDy;
      this.userInput = true;
    }
  }

  processScroll(dy: number): void {
    this.scroll += -dy;
    this.userInput = true;
  }

  /**
   * Moves the controller center to the closest point on a line defined by the camera position and rotation.
   * Adjusts the controller up vector by projecting the current up vector onto the plane defined by the camera right vector.
   */
  resetToCamera(camera: PerspectiveCamera): void {
    const cameraQuat = quat.fromValues(
      camera.rotationQuat[0],
      camera.rotationQuat[1], 
      camera.rotationQuat[2],
      camera.rotationQuat[3]
    );
    
    const invView = mat3.create();
    mat3.fromQuat(invView, cameraQuat);
    mat3.invert(invView, invView);

    const forward = vec3.create();
    vec3.transformMat3(forward, [0, 0, 1], invView);

    const right = vec3.create();
    vec3.transformMat3(right, [1, 0, 0], invView);

    const cameraPos = vec3.fromValues(camera.positionVec[0], camera.positionVec[1], camera.positionVec[2]);
    
    // Move center point
    this.center = closestPoint(cameraPos, forward, this.center);

    // Adjust up vector by projecting it onto the plane defined by the right vector of the camera
    if (this.up) {
      const projection = vec3.create();
      const dot = vec3.dot(this.up, right);
      vec3.scale(projection, right, dot);
      
      const newUp = vec3.create();
      vec3.subtract(newUp, this.up, projection);
      vec3.normalize(newUp, newUp);
      this.up = newUp;
    }
  }

  updateCamera(camera: PerspectiveCamera, dt: number): void {
    const cameraPos = vec3.fromValues(camera.positionVec[0], camera.positionVec[1], camera.positionVec[2]);
    const cameraQuat = quat.fromValues(
      camera.rotationQuat[0],
      camera.rotationQuat[1],
      camera.rotationQuat[2], 
      camera.rotationQuat[3]
    );

    const dir = vec3.create();
    vec3.subtract(dir, cameraPos, this.center);
    const distance = vec3.length(dir);

    // Update distance with scroll (matches Rust: normalize_to((distance.ln() + scroll * dt * 10 * speed).exp()))
    const newDistance = Math.exp(Math.log(distance) + this.scroll * dt * 10.0 * this.speed);
    vec3.normalize(dir, dir);
    vec3.scale(dir, dir, newDistance);

    // Get view transform
    const viewT = mat3.create();
    mat3.fromQuat(viewT, cameraQuat);
    mat3.invert(viewT, viewT);

    const xAxis = vec3.fromValues(viewT[0], viewT[3], viewT[6]);
    const yAxis = this.up || vec3.fromValues(viewT[1], viewT[4], viewT[7]);
    const zAxis = vec3.fromValues(viewT[2], viewT[5], viewT[8]);

    // Calculate offset for panning
    const offset = vec3.create();
    const xOffset = vec3.create();
    const yOffset = vec3.create();
    vec3.scale(xOffset, xAxis, this.shift[1]);
    vec3.scale(yOffset, yAxis, -this.shift[0]);
    vec3.add(offset, xOffset, yOffset);
    vec3.scale(offset, offset, dt * this.speed * 0.1 * distance);

    vec3.add(this.center, this.center, offset);
    vec3.add(cameraPos, cameraPos, offset);

    // Calculate rotation angles
    let theta = this.rotation[0] * dt * this.sensitivity;
    let phi = -this.rotation[1] * dt * this.sensitivity;
    let eta = 0;

    if (this.altPressed) {
      eta = -this.rotation[1] * dt * this.sensitivity;
      theta = 0;
      phi = 0;
    }

    // Create rotation quaternions
    const rotTheta = quat.create();
    const rotPhi = quat.create();
    const rotEta = quat.create();
    
    quat.setAxisAngle(rotTheta, yAxis, theta);
    quat.setAxisAngle(rotPhi, xAxis, phi);
    quat.setAxisAngle(rotEta, zAxis, eta);

    // Combine rotations
    const rot = quat.create();
    quat.multiply(rot, rotTheta, rotPhi);
    quat.multiply(rot, rot, rotEta);

    // Apply rotation to direction
    const newDir = vec3.create();
    vec3.transformQuat(newDir, dir, rot);

    // Check if we're too close to the up vector (gimbal lock prevention)
    const angle = angleShort(yAxis, newDir);
    if (angle < 0.1) {
      vec3.copy(newDir, dir);
    }

    vec3.add(cameraPos, this.center, newDir);

    // Update camera rotation using lookAt
    const negNewDir = vec3.create();
    vec3.negate(negNewDir, newDir);
    vec3.normalize(negNewDir, negNewDir);
    
    // Create lookAt quaternion
    const forward = vec3.create();
    vec3.copy(forward, negNewDir);
    
    const right = vec3.create();
    vec3.cross(right, forward, yAxis);
    vec3.normalize(right, right);
    
    const up = vec3.create();
    vec3.cross(up, right, forward);
    vec3.normalize(up, up);
    
    const rotMat = mat3.fromValues(
      right[0], up[0], -forward[0],
      right[1], up[1], -forward[1],
      right[2], up[2], -forward[2]
    );
    
    const newQuat = quat.create();
    quat.fromMat3(newQuat, rotMat);

    // Update camera properties
    camera.positionVec[0] = cameraPos[0];
    camera.positionVec[1] = cameraPos[1];
    camera.positionVec[2] = cameraPos[2];
    
    camera.rotationQuat[0] = newQuat[0];
    camera.rotationQuat[1] = newQuat[1];
    camera.rotationQuat[2] = newQuat[2];
    camera.rotationQuat[3] = newQuat[3];

    // Apply decay based on fps (60fps reference)
    let decay = Math.pow(0.8, dt * 60.0);
    if (decay < 1e-4) {
      decay = 0.0;
    }

    // Apply decay to inputs
    this.rotation[0] *= decay;
    this.rotation[1] *= decay;
    if (Math.hypot(this.rotation[0], this.rotation[1]) < 1e-4) {
      this.rotation[0] = 0;
      this.rotation[1] = 0;
    }

    this.shift[0] *= decay;
    this.shift[1] *= decay;
    if (Math.hypot(this.shift[0], this.shift[1]) < 1e-4) {
      this.shift[0] = 0;
      this.shift[1] = 0;
    }

    this.scroll *= decay;
    if (Math.abs(this.scroll) < 1e-4) {
      this.scroll = 0;
    }

    this.userInput = false;
  }
}

function closestPoint(orig: vec3, dir: vec3, point: vec3): vec3 {
  const normalizedDir = vec3.create();
  vec3.normalize(normalizedDir, dir);
  
  const lhs = vec3.create();
  vec3.subtract(lhs, point, orig);
  
  const dotP = vec3.dot(lhs, normalizedDir);
  
  const result = vec3.create();
  vec3.scaleAndAdd(result, orig, normalizedDir, dotP);
  return result;
}

function angleShort(a: vec3, b: vec3): number {
  const angle = Math.acos(Math.max(-1, Math.min(1, vec3.dot(a, b) / (vec3.length(a) * vec3.length(b)))));
  if (angle > Math.PI / 2) {
    return Math.PI - angle;
  } else {
    return angle;
  }
}

File: TypescriptWebSplatPort2/web-splat-port/rust-web-splat/src/controller.rs
use cgmath::*;
#[cfg(target_arch = "wasm32")]
use web_time::Duration;
use num_traits::Float;
use std::f32::consts::PI;
#[cfg(not(target_arch = "wasm32"))]
use std::time::Duration;

use winit::keyboard::KeyCode;

use crate::camera::PerspectiveCamera;

#[derive(Debug)]
pub struct CameraController {
    pub center: Point3<f32>,
    pub up: Option<Vector3<f32>>,
    amount: Vector3<f32>,
    shift: Vector2<f32>,
    rotation: Vector3<f32>, // x = yaw input, y = pitch input, z unused (roll removed)
    scroll: f32,
    pub speed: f32,
    pub sensitivity: f32,

    pub left_mouse_pressed: bool,
    pub right_mouse_pressed: bool,
    pub alt_pressed: bool, // no special behavior now, kept for API compatibility
    pub user_inptut: bool,
}

impl CameraController {
    pub fn new(speed: f32, sensitivity: f32) -> Self {
        Self {
            center: Point3::origin(),
            amount: Vector3::zero(),
            shift: Vector2::zero(),
            rotation: Vector3::zero(),
            up: None,
            scroll: 0.0,
            speed,
            sensitivity,
            left_mouse_pressed: false,
            right_mouse_pressed: false,
            alt_pressed: false,
            user_inptut: false,
        }
    }

    pub fn process_keyboard(&mut self, key: KeyCode, pressed: bool) -> bool {
        let amount = if pressed { 1.0 } else { 0.0 };
        let processed = match key {
            KeyCode::KeyW | KeyCode::ArrowUp => {
                self.amount.z += amount; // forward
                true
            }
            KeyCode::KeyS | KeyCode::ArrowDown => {
                self.amount.z += -amount; // back
                true
            }
            KeyCode::KeyA | KeyCode::ArrowLeft => {
                self.amount.x += -amount; // left
                true
            }
            KeyCode::KeyD | KeyCode::ArrowRight => {
                self.amount.x += amount; // right
                true
            }

            // Roll removed: Q/E no longer affect rotation.z. If you prefer, you can
            // repurpose these to something else later. For now, ignore.
            KeyCode::KeyQ | KeyCode::KeyE => false,

            KeyCode::Space => {
                self.amount.y += amount; // up
                true
            }
            KeyCode::ShiftLeft => {
                self.amount.y += -amount; // down
                true
            }
            _ => false,
        };
        self.user_inptut = processed;
        processed
    }

    pub fn process_mouse(&mut self, mouse_dx: f32, mouse_dy: f32) {
        if self.left_mouse_pressed {
            // Left-drag: yaw (around up) and pitch (around right)
            self.rotation.x += mouse_dx as f32; // yaw input
            self.rotation.y += mouse_dy as f32; // pitch input
            self.user_inptut = true;
        }
        if self.right_mouse_pressed {
            // Right-drag: pan
            self.shift.y += -mouse_dx as f32;
            self.shift.x += mouse_dy as f32;
            self.user_inptut = true;
        }
    }

    pub fn process_scroll(&mut self, dy: f32) {
        self.scroll += -dy;
        self.user_inptut = true;
    }

    /// moves the controller center to the closest point on a line defined by the camera position and rotation
    /// ajusts the controller up vector by projecting the current up vector onto the plane defined by the camera right vector
    pub fn reset_to_camera(&mut self, camera: PerspectiveCamera) {
        let inv_view = camera.rotation.invert();
        let forward = inv_view * Vector3::unit_z();
        let right = inv_view * Vector3::unit_x();

        // move center point
        self.center = closest_point(camera.position, forward, self.center);
        // adjust up vector by projecting it onto the plane defined by the right vector of the camera
        if let Some(up) = &self.up {
            let new_up = up - up.project_on(right);
            self.up = Some(new_up.normalize());
        }
    }

    pub fn update_camera(&mut self, camera: &mut PerspectiveCamera, dt: Duration) {
        let dt: f32 = dt.as_secs_f32();

        // radial dolly with logarithmic distance mapping
        let mut dir = camera.position - self.center;
        let distance = dir.magnitude();
        dir = dir.normalize_to((distance.ln() + self.scroll * dt * 10.0 * self.speed).exp());

        // camera local axes in world space
        let view_t: Matrix3<f32> = camera.rotation.invert().into();
        let x_axis = view_t.x;                       // camera right
        let y_axis = self.up.unwrap_or(view_t.y);    // world/camera up
        // let z_axis = view_t.z;                    // camera forward (unused for roll)

        // pan (screen-space)
        let offset = (self.shift.y * x_axis - self.shift.x * y_axis) * dt * self.speed * 0.1 * distance;
        self.center += offset;
        camera.position += offset;

        // convert smoothed mouse deltas into angular velocities
        let theta = Rad((self.rotation.x) * dt * self.sensitivity);  // yaw around up
        let phi   = Rad((-self.rotation.y) * dt * self.sensitivity); // pitch around right

        // roll removed: no eta around z_axis, and alt no longer changes mode
        let rot_yaw   = Quaternion::from_axis_angle(y_axis, theta);
        let rot_pitch = Quaternion::from_axis_angle(x_axis, phi);
        let rot = rot_yaw * rot_pitch;

        // apply yaw+pitch to the camera-to-center vector
        let mut new_dir = rot.rotate_vector(dir);

        // prevent flipping over the poles: if too close to up direction, keep the old dir
        // (you can tighten/loosen 0.1 rad as needed)
        if angle_short(y_axis, new_dir) < Rad(0.1) {
            new_dir = dir;
        }

        camera.position = self.center + new_dir;

        // face the center with the chosen up axis (keeps horizon level; no roll)
        camera.rotation = Quaternion::look_at(-new_dir, y_axis);

        // decay based on fps
        let mut decay = (0.8).powf(dt * 60.0);
        if decay < 1e-4 {
            decay = 0.0;
        }

        // decay inputs
        self.rotation.x *= decay; // yaw
        self.rotation.y *= decay; // pitch
        self.rotation.z = 0.0;    // ensure roll stays zero
        if self.rotation.truncate().magnitude() < 1e-4 {
            self.rotation = Vector3::zero();
        }

        self.shift *= decay;
        if self.shift.magnitude() < 1e-4 {
            self.shift = Vector2::zero();
        }

        self.scroll *= decay;
        if self.scroll.abs() < 1e-4 {
            self.scroll = 0.0;
        }

        self.user_inptut = false;
    }
}

fn closest_point(orig: Point3<f32>, dir: Vector3<f32>, point: Point3<f32>) -> Point3<f32> {
    let dir = dir.normalize();
    let lhs = point - orig;
    let dot_p = lhs.dot(dir);
    orig + dir * dot_p
}

fn angle_short(a: Vector3<f32>, b: Vector3<f32>) -> Rad<f32> {
    let angle = a.angle(b);
    if angle > Rad(PI / 2.0) {
        Rad(PI) - angle
    } else {
        angle
    }
}
